clang/test/CodeGen/aie/aie2/aie2-bank-annotation.cpp

```diff
@@ -0,0 +1,161 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 4
+//===- aie2-bank-annotation.cpp----------------------------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+// Test to check the bank annotation for AIE2, specifically the translation of __aie_dm_resource_* to the correct address space.
+// Test expression where pointer with AS is assigned to pointer with Default AS.
+// RUN: %clang -O1 --target=aie2 -S -emit-llvm %s -o - | FileCheck %s
+
+// CHECK-LABEL: define dso_local noundef i32 @_Z6squarePU3AS5iS0_(
+// CHECK-SAME: ptr addrspace(5) nocapture readonly [[NUM:%.*]], ptr addrspace(5) nocapture writeonly [[MEM:%.*]]) local_unnamed_addr #[[ATTR0:[0-9]+]] {
+// CHECK-NEXT:  entry:
+// CHECK-NEXT:    [[TMP0:%.*]] = load i32, ptr addrspace(5) [[NUM]], align 4, !tbaa [[TBAA2:![0-9]+]]
+// CHECK-NEXT:    [[MUL:%.*]] = mul nsw i32 [[TMP0]], [[TMP0]]
+// CHECK-NEXT:    store i32 [[MUL]], ptr addrspace(5) [[MEM]], align 4, !tbaa [[TBAA2]]
+// CHECK-NEXT:    ret i32 [[MUL]]
+//
+int square(int __aie_dm_resource_a *num, int __aie_dm_resource_a *mem) {
+    *mem = (*num) * (*num);
+    return *mem;
+}
+
+// CHECK-LABEL: define dso_local noundef i32 @_Z6squarePU3AS6iS0_(
+// CHECK-SAME: ptr addrspace(6) nocapture readonly [[NUM:%.*]], ptr addrspace(6) nocapture writeonly [[MEM:%.*]]) local_unnamed_addr #[[ATTR0]] {
+// CHECK-NEXT:  entry:
+// CHECK-NEXT:    [[TMP0:%.*]] = load i32, ptr addrspace(6) [[NUM]], align 4, !tbaa [[TBAA2]]
+// CHECK-NEXT:    [[MUL:%.*]] = mul nsw i32 [[TMP0]], [[TMP0]]
+// CHECK-NEXT:    store i32 [[MUL]], ptr addrspace(6) [[MEM]], align 4, !tbaa [[TBAA2]]
+// CHECK-NEXT:    ret i32 [[MUL]]
+//
+int square(int __aie_dm_resource_b *num, int __aie_dm_resource_b *mem) {
+    *mem = (*num) * (*num);
+    return *mem;
+}
+
+
+// CHECK-LABEL: define dso_local noundef i32 @_Z6squarePU3AS6iPi(
+// CHECK-SAME: ptr addrspace(6) nocapture readonly [[NUM:%.*]], ptr nocapture writeonly [[MEM:%.*]]) local_unnamed_addr #[[ATTR0]] {
+// CHECK-NEXT:  entry:
+// CHECK-NEXT:    [[TMP0:%.*]] = load i32, ptr addrspace(6) [[NUM]], align 4, !tbaa [[TBAA2]]
+// CHECK-NEXT:    [[MUL:%.*]] = mul nsw i32 [[TMP0]], [[TMP0]]
+// CHECK-NEXT:    store i32 [[MUL]], ptr [[MEM]], align 4, !tbaa [[TBAA2]]
+// CHECK-NEXT:    ret i32 [[MUL]]
+//
+int square(int __aie_dm_resource_b *num, int *mem) {
+    *mem = (*num) * (*num);
+    return *mem;
+}
+
+
+// CHECK-LABEL: define dso_local noundef i32 @_Z6squarePU3AS7i(
```

Comment:
I think those "simple" tests below should be in a separate file where we really just test the translation of memory banks into address spaces. That file can then be added as a test in the first commit.

The tests above should remain here in the second commit, as they test that implicit address space casts are allowed.